nixos/modules/services/networking/tailscale.nix

```diff
@@ -60,6 +61,8 @@ in {
         ''"FLAGS=--tun ${lib.escapeShellArg cfg.interfaceName}"''
       ] ++ (lib.optionals (cfg.permitCertUid != null) [
         "TS_PERMIT_CERT_UID=${cfg.permitCertUid}"
+      ]) ++ (lib.optionals (httpsProxy != null) [
+        "HTTPS_PROXY=${httpsProxy}"
```

Comment:
We also need to propagate HTTP_PROXY (some of tailscaled's outbound connections are Noise over cleartext HTTP) and NO_PROXY (so things like connecting to localhost still work).

You also probably want to pull the values from config.networking.proxy.envVars.{no_proxy, http_proxy, https_proxy}, because at least no_proxy gets some default value treatment between the user-specified config and the final value.

Reply:
Tailscale agent code only honors HTTPS_PROXY variable. It's only used to establish control connection with tailscale.io over https.

Reply:
That's not correct. The Tailscale 2021 control protocol uses Noise over plaintext HTTP, and only downgrades to Noise over TLS if the HTTP connection fails. Noise over TLS is much less efficient because it results in double-encryption of every byte, and more RTTs to establish the connection.

https://github.com/tailscale/tailscale/blob/main/control/controlhttp/client.go#L264 is the dial plan for control plane connections. See how it first attempts an HTTP connection before falling through to HTTPS.
If you trace the proxy functions all the way down, you'll see that on NixOS it'll end up using https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/net/http/transport.go;l=439 from net/http, and if you trace _that_ implementation further, you'll see that it uses HTTP_PROXY when dialing http:// URLs: https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/vendor/golang.org/x/net/http/httpproxy/proxy.go;l=117;drc=38cfb3be9d486833456276777155980d1ec0823e;bpv=1;bpt=1

Please pass through all vars I listed above from config.networking.proxy.envVars, Go's HTTP proxy logic obeys all three and therefore Tailscale's dialing logic.